### PR TITLE
SQL-1059: Add error handling to catch any unexpected panic

### PR DESCRIPTION
**Pull request by bucaojit**
Added `panic_safe_exec` macro that executes the passed function and captures panics that occur. 
Used the macro in the  `extern "C" fn SQLXXX ` functions where functions are implemented, and functions with the `unimplemented` macro except for the functions with a descriptor handle as MongoHandleRef::from() is not implemented for descriptor handles.

### CODE REVIEW

File: odbc/src/api/driver_connect_tests.rs [hunk 1]
```diff
@@ -1,142 +0,0 @@
-macro_rules! test_connection_diagnostics {
```
Comment (bucaojit) [inline odbc/src/api/driver_connect_tests.rs:1]: This file was initially deleted in SQL-1036 and was accidentally recreated after a git merge.  Deleting it again here.

File: odbc/src/api/functions.rs [hunk 2]
```diff
@@ -256,8 +300,8 @@ pub unsafe extern "C" fn SQLBulkOperations(
 /// Because this is a C-infereface, this is necessarily unsafe
 ///
 #[no_mangle]
-pub unsafe extern "C" fn SQLCancel(_statement_handle: HStmt) -> SqlReturn {
-    unimplemented!()
+pub unsafe extern "C" fn SQLCancel(statement_handle: HStmt) -> SqlReturn {
+    panic_safe_exec!(|| { unimplemented!() }, statement_handle);
```
Comment (pmeredit) [inline odbc/src/api/functions.rs:304]: [nit] We could define another macro call `unimpl` (or similar) that does this line.

**Review by pmeredit — APPROVED** [review]
LGTM

**Comment by bucaojit** [comment]
evergreen merge

**Pull request merged**